quic/s2n-quic-core/src/sync/spsc/slice.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{cell::UnsafeCell, mem::MaybeUninit, ops::Deref};
+
+#[repr(transparent)]
+pub struct Cell<T>(MaybeUninit<UnsafeCell<T>>);
+
+impl<T> Cell<T> {
+    #[inline]
+    pub unsafe fn write(&self, value: T) {
+        UnsafeCell::raw_get(self.0.as_ptr()).write(value);
+    }
+
+    #[inline]
+    pub unsafe fn take(&self) -> T {
+        self.0.assume_init_ref().get().read()
+    }
+}
+
+#[derive(Debug)]
+pub struct Slice<'a, T>(pub(super) &'a [T]);
+
+impl<'a, T> Slice<'a, Cell<T>> {
+    #[inline]
+    pub unsafe fn assume_init(self) -> Slice<'a, UnsafeCell<T>> {
+        Slice(&*(self.0 as *const [Cell<T>] as *const [UnsafeCell<T>]))
+    }
+
+    #[inline]
+    pub unsafe fn write(&self, index: usize, value: T) {
+        self.0.get_unchecked(index).write(value)
+    }
+
+    #[inline]
+    pub unsafe fn take(&self, index: usize) -> T {
+        self.0.get_unchecked(index).take()
+    }
+}
+
+impl<'a, T> Slice<'a, UnsafeCell<T>> {
+    #[inline]
+    pub unsafe fn into_mut(self) -> &'a mut [T] {
+        let ptr = self.0.as_ptr() as *mut T;
+        let len = self.0.len();
+        core::slice::from_raw_parts_mut(ptr, len)
+    }
+}
+
+impl<'a, T> Deref for Slice<'a, T> {
+    type Target = [T];
+
+    #[inline]
+    fn deref(&self) -> &[T] {
+        self.0
+    }
+}
+
+impl<'a, T: PartialEq> PartialEq<[T]> for Slice<'a, UnsafeCell<T>> {
+    #[inline]
+    fn eq(&self, other: &[T]) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+
+        for (a, b) in self.iter().zip(other) {
+            if unsafe { &*a.get() } != b {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<'a, T: PartialEq> PartialEq<Slice<'a, UnsafeCell<T>>> for [T] {
+    #[inline]
+    fn eq(&self, other: &Slice<'a, UnsafeCell<T>>) -> bool {
+        other.eq(self)
+    }
+}
+
+impl<'a, T: PartialEq> PartialEq<Slice<'a, UnsafeCell<T>>> for &[T] {
+    #[inline]
+    fn eq(&self, other: &Slice<'a, UnsafeCell<T>>) -> bool {
+        other.eq(self)
+    }
+}
+
+#[derive(Debug)]
+pub struct Pair<S> {
+    pub head: S,
+    pub tail: S,
+}
+
+impl<'a, T> Pair<Slice<'a, Cell<T>>> {
+    #[inline]
+    pub unsafe fn assume_init(self) -> Pair<Slice<'a, UnsafeCell<T>>> {
+        Pair {
+            head: self.head.assume_init(),
+            tail: self.tail.assume_init(),
+        }
+    }
+
+    #[inline]
+    pub unsafe fn write(&self, index: usize, value: T) {
+        self.cell(index).write(value)
+    }
+
+    #[inline]
+    pub unsafe fn take(&self, index: usize) -> T {
+        self.cell(index).take()
+    }
+
+    unsafe fn cell(&self, index: usize) -> &Cell<T> {
+        if let Some(cell) = self.head.0.get(index) {
+            cell
+        } else {
+            assume!(index >= self.head.0.len());
```

Comment:
They're they exact same when debug assertions are enabled. However, when they're not, the `assume` tells the compiler that it's impossible for this check to be false and then the compiler will optimize the code based on that assumption. Without extensive testing/proofs it's generally best to stick with a debug_assert. But in this case, I'm using `loom`, `kani`, and `bolero` so the confidence is quite high that these assumptions always hold.

Reply:
I guess my question is more, is there something about this code that makes you believe the compiler would benefit from the `assume!` and give us better performance, or is it more that we might as well use `assume!` since all the testing and proofs give us high confidence that its safe.

Reply:
Yeah, it's mostly the latter. I want to try and give the compiler as much information as we can about the assumptions and prove that they will always hold. In this particular case, I didn't dive into the asm to know if it made a difference in the final outcome.